kafkabp/consumer.go

```diff
@@ -0,0 +1,222 @@
+	return c
+}
+
+func (kc *consumer) getPartitions() []int32 {
+	p, _ := kc.partitions.Load().([]int32)
+	return p
+}
+
+func (kc *consumer) getPartitionConsumers() []sarama.PartitionConsumer {
+	pc, _ := kc.partitionConsumers.Load().([]sarama.PartitionConsumer)
+	return pc
+}
+
+// reset recreates the consumer and assigns partitions.
+func (kc *consumer) reset() error {
+	if c := kc.getConsumer(); c != nil {
+		if err := c.Close(); err != nil {
+			kc.cfg.Logger.Log(context.Background(), "kafkabp.consumer.reset: Error closing the consumer:"+err.Error())
+		}
+	}
+
+	rebalance := func() error {
+		c, err := sarama.NewConsumer(kc.cfg.Brokers, kc.sc)
+		if err != nil {
+			return err
+		}
+
+		partitions, err := c.Partitions(kc.cfg.Topic)
+		if err != nil {
+			return err
+		}
+
+		kc.consumer.Store(c)
+		kc.partitions.Store(partitions)
+		return nil
+	}
+
+	err := rebalance()
+	if err != nil {
+		metricsbp.M.Counter("kafka.consumer.rebalance.failure").Add(1)
+		return err
+	}
+
+	metricsbp.M.Counter("kafka.consumer.rebalance.success").Add(1)
+	return nil
+}
+
+// Close closes all partition consumers first, then the parent consumer.
+func (kc *consumer) Close() error {
+	// Return early if closing is already in progress
+	if !atomic.CompareAndSwapInt64(&kc.closed, 0, 1) {
+		return nil
+	}
+
+	partitionConsumers := kc.getPartitionConsumers()
+	for _, pc := range partitionConsumers {
+		// leaves room to drain pc's message and error channels
+		pc.AsyncClose()
+	}
+	// wait for the Consume function to return
+	kc.wg.Wait()
+	return kc.getConsumer().Close()
+}
+
+// Consume consumes Kafka messages and errors from each partition's consumer.
+// It is necessary to call Close() on the KafkaConsumer instance once all
+// operations are done with the consumer instance.
+func (kc *consumer) Consume(
+	messagesFunc ConsumeMessageFunc,
+	errorsFunc ConsumeErrorFunc,
+) error {
+	defer atomic.StoreInt64(&kc.consumeReturned, 1)
+	kc.wg.Add(1)
+	defer kc.wg.Done()
+
+	// Sarama could close the channels (and cause the goroutines to finish) in
+	// two cases, where we want different behavior:
+	//   - in case of partition rebalance: restart goroutines
+	//   - in case of call to Close/AsyncClose: exit
+	var wg sync.WaitGroup
+	for {
+		// create a partition consumer for each partition
+		consumer := kc.getConsumer()
+		partitions := kc.getPartitions()
+		partitionConsumers := make([]sarama.PartitionConsumer, 0, len(partitions))
+
+		for _, p := range partitions {
+			partitionConsumer, err := consumer.ConsumePartition(kc.cfg.Topic, p, kc.offset)
+			if err != nil {
+				return err
+			}
+			partitionConsumers = append(partitionConsumers, partitionConsumer) // for closing individual partitions when Close() is called
+
+			// consume partition consumer messages
+			wg.Add(1)
+			go func(pc sarama.PartitionConsumer) {
+				defer wg.Done()
+				for m := range pc.Messages() {
+					// Wrap in anonymous function for easier defer.
+					func() {
```

Comment:
:no_bell: At this point we're fairly deeply intended, I wonder if moving some of this into methods will improve the cognitive complexity